users/test/test_views.py

```diff
@@ -5,16 +5,105 @@
+        self.factory = APIRequestFactory()
+
+    def test_unauthorized_user_follow(self) -> None:
+        url = reverse("user-follow", kwargs={"pk": self.user_one.id})
+        response = self.client.post(
+            url, kwargs={"pk": self.user_one.id}, format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+    def test_unauthorized_user_unfollow(self) -> None:
+        url = reverse("user-follow", kwargs={"pk": self.user_one.id})
+        response = self.client.delete(
+            url, kwargs={"pk": self.user_one.id}, format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+    def test_authorized_user_follow(self) -> None:
+        url = reverse("user-follow", kwargs={"pk": self.user_one.id})
+        request = self.factory.post(url)
+        force_authenticate(request, user=self.user_two)
+        response = follow_view(request, pk=self.user_one.id)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        response.render()
+        self.assertContains(response, text=self.user_one.username)
+
+    def test_authorized_user_unfollow(self) -> None:
+        url = reverse("user-follow", kwargs={"pk": self.user_one.id})
+        request = self.factory.delete(url)
+        force_authenticate(request, user=self.user_two)
+        response = follow_view(request, pk=self.user_one.id)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_authorized_get_followers(self) -> None:
+        factory = APIRequestFactory()
+        url = reverse("user-follow", kwargs={"pk": self.user_one.id})
+        request = factory.get(url)
+        force_authenticate(request, user=self.user_two)
+        response = follow_view(request, pk=self.user_one.id)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_user_cannot_follow_self(self) -> None:
+        url = reverse("user-follow", kwargs={"pk": self.user_one.id})
+        request = self.factory.post(url)
+        force_authenticate(request, user=self.user_one)
+        response = follow_view(request, pk=self.user_one.id)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        response.render()
+        self.assertEqual(
+            response.data, {"details": "you cannot follow yourself"}
+        )
+
+    def test_user_cannot_follow_non_existent_user(self) -> None:
+        url = reverse("user-follow", kwargs={"pk": self.user_one.id})
+        request = self.factory.post(url)
+        force_authenticate(request, user=self.user_one)
+        response = follow_view(request, pk=self.user_one.id + 100)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_user_cannot_follow_same_user(self) -> None:
+        url = reverse("user-follow", kwargs={"pk": self.user_two.id})
+        request = self.factory.post(url)
+        force_authenticate(request, user=self.user_one)
```

Comment:
try to immitate how the user will be using the system eg not force authenticating but passing the token in the headers

Reply:
Noted